bin/ch-completion.bash

```diff
@@ -0,0 +1,403 @@
+#     see https://github.com/git/git/tree/master/contrib/completion
+#   * Add support for mid-line completion (possibly using COMP_POINT,
+#     see https://devmanual.gentoo.org/tasks-reference/completion/index.html).
+#
+
+# Minimum supported bash version. According to
+# http://mywiki.wooledge.org/BashFAQ/061 and my own testing, negative array
+# indexing was introduced in this version. To simplify confusing syntax, we this
+# script is unsupported for bash < 4.2.0.
+bash_vmin="4.2.0"
+
+# Check bash version
+bash_v=$(bash --version | head -1 | grep -Eo "[0-9\.]{2,}[0-9]")
+if [[ $(printf "%s\n%s\n" "$bash_vmin" "$bash_v" | sort -V) < "$bash_vmin" ]]; then
+    echo "ch-completion.bash: unsupported bash version ($bash_v < $bash_vmin)"
+    return
+fi
+
+# Debugging log
+if [[ -f "/tmp/ch-completion.log" && -n "$CH_COMPLETION_DEBUG" ]]; then
+    rm /tmp/ch-completion.log
+    printf "completion log\n\n" >> /tmp/ch-completion.log
+fi
+
+## ch-image ##
+
+# Subcommands and options for ch-image
+#
+_image_build_opts="-b --bind --build-arg -f --file --force
+                   --force-cmd -n --dry-run --parse-only -t --tag"
+
+_image_common_opts="-a --arch --always-download --auth --cache
+                    --cache-large --dependencies -h --help
+                    --no-cache --no-lock --profile --rebuild
+                    --password-many -s --storage --tls-no-verify
+                    -v --verbose --version"
+
+_image_subcommands="build build-cache delete gestalt
+                    import list pull push reset undelete"
+
+# archs taken from ARCH_MAP in charliecloud.py
+_archs="amd64 arm/v5 arm/v6 arm/v7 arm64/v8 386 mips64le ppc64le s390x"
+
+## ch-image ##
+
+# Completion function for ch-image
+#
+_ch_image_complete () {
+    local prev
+    local cur
+    local words
+    local sub_cmd
+    local strg_dir
+    local extras=
+    _get_comp_words_by_ref -n : cur prev words
+
+    # To find the subcommand and storage directory, we pass the associated
+    # functions the current command line minus the last word (note that
+    # “${words[@]::${#words[@]}-1}” in bash is analagous to “words[:-1]” in
+    # python). We do this because the last word is most likely either an empty
+    # string, or is not yet complete. We don't lose anything by dropping the
+    # empty string, and an incomplete word in the command line can lead to false
+    # positives from these functions and consequently unexpected behavior, so we
+    # don't consider it.
+    sub_cmd=$(_ch_image_subcmd_get "${words[@]::${#words[@]}-1}")
+    strg_dir=$(_ch_find_storage "${words[@]::${#words[@]}-1}")
+
+    # Populate debug log
+    if [[ -n "$CH_COMPLETION_DEBUG" ]]; then
+        # shellcheck disable=SC2129
+        echo "\$ ${words[*]}" >> /tmp/ch-completion.log
+        echo "  current: $cur" >> /tmp/ch-completion.log
+        echo "  previous: $prev" >> /tmp/ch-completion.log
+        echo "  sub command: $sub_cmd" >> /tmp/ch-completion.log
+    fi
```

Comment:
Probably best to encapsulate the `$CH_COMPLETION_DEBUG` test in a logging function.